util/exercise/src/structs.rs

```diff
@@ -0,0 +1,159 @@
+//! Rust struct for canonical test data.
+//!
+//! See https://github.com/exercism/problem-specifications/blob/master/canonical-schema.json
+//! for more details on the JSON schema, which makes it possible to implement
+//! `serde::Deserialize`.
+
+use crate::{self as exercise};
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+
+macro_rules! pub_struct_serde_getters {
+    (
+        pub struct $name:ident {
+            $( $fname:ident : $ftype:ty ),* $(,)*
+        }
+    ) => {
+        #[derive(Serialize, Deserialize, Debug)]
+        pub struct $name {
+            $($fname : $ftype),*
+        }
+
+        impl $name {
+            $(
+            pub fn $fname(&self) -> &$ftype {
+                &self.$fname
+            }
+            )*
+        }
+    };
+}
+
+pub_struct_serde_getters! {
+    pub struct CanonicalData {
+        exercise: Exercise,
+        version: Version,
+        comments: Option<Comments>,
+        cases: TestGroup,
+    }
+}
+
+type Exercise = String;
+type Version = String;
+type Comments = Vec<String>;
+type TestGroup = Vec<LabeledTestItem>;
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum LabeledTestItem {
+    Single(LabeledTest),
+    Array(LabeledTestGroup),
+}
+
+pub_struct_serde_getters! {
+    pub struct LabeledTest {
+        description: Description,
+        optional: Option<Optional>,
+        comments: Option<Comments>,
+        property: Property,
+        input: Input,
+        expected: Expected,
+    }
+}
+
+impl LabeledTest {
+    pub fn generate_test_function(&self) -> String {
```

Comment:
The templates approach feels like a very good idea. Honestly, the `generate_test_function` method was about not breaking the previous behaviour; I should say mostly not breaking, as the unwrapping is a bit different (also `use_maplit`).

One interesting aspect with `syn` is that generating `ItemFn` is straightforward using `parse_str` (or even a full source `File`). And `ItemFn` implements `Eq` and `Hash`, which should make finding new, updated and deleted versions of the tests relatively straightforward.

That being said, I'm not very familiar with either `handlebars` or `tera`, and they may offer an even simpler way to compare old vs new than going through `syn`. Is that the case?

And they could probably deliver more readable code; `quote` is great, but, absent `rustfmt`, it does not deliver the most readable code from `syn`. :smile:

Reply:
Looking back on this, implementing `generate_test_function` doesn't accomplish much; I had high hopes for the `Data` enum, but it hasn't lived up to it.

As it is, the final output is slightly different from what it used to be. So, this is one change I'm quite willing to revert to the original implementation in `lib.rs` (with the required changes to use `CanonicalData` instead). This would probably be for the best until we can wrestle the templates. What do you think?

Reply:
Yeah, I think that until we can get some proper templates in, I think I prefer the string-formatting "templates" over the imperative version. It just makes it easier, for me at least, to visualize what the output will be.

Reply:
I agree. I reverted to the former `generate_test_function` function, although this required some minor modifications to use the public fields.

This still uses the `Data::to_literal` method, since it works fine, but I'm thinking this could also be reverted to the former `into_literal` function, by replacing the `input` and `expected` type with `Value`s; one modification would be to change the output of the function from a `Result<String>` to a `String`, since it can't fail as long as deserialization has succeeded (and eliminating the propagation to callers). In that case, I think resetting to an earlier commit (916d26b?) might make for a cleaner history though.